lib/sidekiq/job/iterable.rb

```diff
@@ -0,0 +1,261 @@
+      #
+      def on_resume
+      end
+
+      # A hook to override that will be called each time the job is interrupted.
+      #
+      # This can be due to throttling, `max_job_runtime` configuration,
+      # or sidekiq stopping.
+      #
+      def on_shutdown
+      end
+
+      # A hook to override that will be called when the job finished iterating.
+      #
+      def on_complete
+      end
+
+      # The enumerator to be iterated over.
+      #
+      # @return [Enumerator]
+      #
+      # @raise [NotImplementedError] with a message advising subclasses to
+      #     implement an override for this method.
+      #
+      def build_enumerator(*)
+        raise NotImplementedError, "#{self.class.name} must implement a '#build_enumerator' method"
+      end
+
+      # The action to be performed on each item from the enumerator.
+      #
+      # @return [void]
+      #
+      # @raise [NotImplementedError] with a message advising subclasses to
+      #     implement an override for this method.
+      #
+      def each_iteration(*)
+        raise NotImplementedError, "#{self.class.name} must implement an '#each_iteration' method"
+      end
+
+      # A hook to override that can be used to throttle a job when a given condition is met.
+      #
+      # If a job is throttled, it will be interrupted and retried after a backoff period
+      # (0 by default) has passed.
+      #
+      # The backoff can be configured via `sidekiq_options` per job:
+      #
+      #   sidekiq_options iteration: { retry_backoff: 30 } # in seconds
+      #
+      #  or globally:
+      #
+      #   Sidekiq::Config::DEFAULTS[:iteration][:retry_backoff] = 30 # in seconds
+      #
+      # @return [Boolean]
+      #
+      def throttle?
+        false
+      end
+
+      # @api private
+      def perform(*arguments)
+        @arguments = arguments
+        fetch_previous_iteration_state
+
+        @executions += 1
+        @start_time = Time.now
+
+        enumerator = build_enumerator(*arguments, cursor: @cursor)
+        unless enumerator
+          logger.debug("'#build_enumerator' returned nil, skipping the job.")
+          return
+        end
+
+        assert_enumerator!(enumerator)
+
+        if @executions == 1
+          on_start
+        else
+          on_resume
+        end
+
+        completed = catch(:abort) do
+          iterate_with_enumerator(enumerator, arguments)
+        end
+
+        on_shutdown
+        completed = handle_completed(completed)
+
+        if completed
+          on_complete
+
+          logger.info {
+            format("Completed iterating. times_interrupted=%d total_time=%.3f", @times_interrupted, @total_time)
+          }
+        else
+          reenqueue_iteration_job
+        end
+      end
+
+      private
+
+      def fetch_previous_iteration_state
+        state = Sidekiq.redis { |conn| conn.get("it-#{jid}") }
+
+        if state
+          state_hash = Sidekiq.load_json(state)
```

Comment:
Should we handle errors here?

Reply:
No, exceptional cases should generate exceptions. I do think you should use native Redis hashes instead of JSON, so you'd use `hgetall` rather than `get`.

Reply:
We currently store the iteration cursor as part of this metadata. When using json, we get the correct data type when loading from json. But with redis hashes we will always get strings as values and the user will need to remember to manually cast the string value to the correct datatype. When people will use some custom enumerators or we start supporting custom columns for active record enumerator in the future, we will get other types of cursors, other than integers, and this will be a problem.

Wdyt?

Reply:
You can make the cursor JSON if it is user-supplied but Sidekiq's internal iteration data can be stored element-by-element in the hash and coerced in Ruby. JSON is meant to be used for interoperability but for our internal data, I'd rather the data be directly addressable in Redis, rather than requiring deserialization.